src/proxy/http/profiles.rs

```diff
@@ -392,19 +472,30 @@ pub mod router {
         }
     }
 
-    impl<T, G, M, R, B, RMk, RSvc> svc::Service<T> for MakeSvc<G, M, R, B>
+    impl<T, G, M, R, RBody, MBody, MSvc, RMk, RSvc> svc::Service<T> for MakeSvc<G, M, R, RBody, MBody>
     where
-        T: CanGetDestination + WithRoute + Clone,
+        T: CanGetDestination + WithRoute + WithAddr + Eq + Hash + Clone,
```

Comment:
I know this occurs a few places in the proxy code, but seeing this `where` makes me think we should try and have a comment explaining the difference at least between the the `R..` and non-`R..` type bounds. It wasn't very obvious to me when first working on this area of the code.

Reply:
Changing the type parameters from single letters to whole words makes the bounds _longer_, but WDYT of something like this:
```rust
    impl<T, Routes, InnerMk, RouteLayer, RouteBody, InnerBody, InnerSvc, RouteMk, RouteSvc> svc::Service<T> 
        for MakeSvc<Routes, InnerMk, RouteLayer, RouteBody, InnerBody>
    where
        T: CanGetDestination + WithRoute + WithAddr + Eq + Hash + Clone,
        <T as WithRoute>::Output: Eq + Hash + Clone,
        InnerMk: rt::Make<T, Value = InnerSvc> + Clone,
        InnerSvc: svc::Service<http::Request<InnerBody>> + Clone,
        InnerSvc::Error: Into<Error>,
        Routes: GetRoutes,
        RouteLayer: svc::Layer<svc::shared::Shared<ConcreteRouter<T, InnerMk::Value, InnerBody>>, Service = RouteMk>
            + Clone,
        RouteMk: rt::Make<<T as WithRoute>::Output, Value = RouteSvc> + Clone,
        RouteSvc: svc::Service<http::Request<RouteBody>> + Clone,
        RouteSvc::Error: Into<Error>,
    {
```

In particular, I think making the two "families" of types all share a common prefix might make this a little easier to understand

Reply:
I still think longer type parameter names could be helpful for the short term, but don't like we need to block on it!